File: python/parmana/crypto/offline_verifier.py
"""
Standalone, offline verification of an Execution Trust Record.

Python counterpart to packages/crypto/src/OfflineVerifier.ts (PQC
audit RED-1, docs/VERIFICATION-GAPS.md) -- proves the canonical
serialization algorithm and Ed25519 verification are independently
reproducible in a second language, not just documented as "should be
portable." Zero network calls, zero disk reads (beyond files the
caller explicitly passes in), zero environment variables.

Large records (over 4096 bytes of canonical content) may have been
signed by AWS KMS as a fixed size commitment, because KMS caps a raw
Ed25519 message at 4096 bytes. This module accepts a raw signature for
any record, and additionally accepts the commitment form for a record
over that limit (see docs/adr/ADR-0010). A commitment signature over a
record at or below the limit is rejected.

ML-DSA-65 (Dilithium3) verification is intentionally out of scope
here: the `cryptography` package version this SDK currently depends
on does not yet expose `cryptography.hazmat.primitives.asymmetric.
ml_dsa`. A hybrid-signed record's `signatures` array cannot be
independently checked from Python until that support lands upstream
and this module is updated to use it -- stated plainly rather than
silently skipped or faked.
"""

from __future__ import annotations

import hashlib
from dataclasses import dataclass, field
from typing import Any

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey
from cryptography.hazmat.primitives.hashes import SHA256, Hash
from cryptography.hazmat.primitives.serialization import load_pem_public_key

from .canonical import canonical_serialize

_SUPPORTED_ALGORITHMS = {"ed25519"}

# Large message commitment (v1), mirroring packages/crypto/src/
# SignatureCommitment.ts. AWS KMS caps a raw Ed25519 message at 4096
# bytes, so the KMS signer signs a longer message as a fixed size
# commitment: the prefix below followed by the SHA-512 digest of the
# message. The scheme depends only on message length, so no marker is
# stored on the record. A message at or below the limit must verify
# raw. A longer message verifies either raw (a local key signed it
# directly) or as a commitment. A commitment signature over a message at
# or below the limit is not accepted.
_KMS_RAW_MESSAGE_LIMIT_BYTES = 4096
_COMMITMENT_PREFIX = b"PARMANA-ED25519-LARGE-MESSAGE-V1\x00"


def _commitment_message(message: bytes) -> bytes:
    return _COMMITMENT_PREFIX + hashlib.sha512(message).digest()


def _verify_ed25519_with_commitment(
    public_key: Ed25519PublicKey, signature: bytes, message: bytes
) -> None:
    """Raises InvalidSignature unless the signature is valid for `message`."""

    try:
        public_key.verify(signature, message)
        return
    except InvalidSignature:
        if len(message) <= _KMS_RAW_MESSAGE_LIMIT_BYTES:
            raise

    public_key.verify(signature, _commitment_message(message))


def _canonical_execution_trust_record(record: dict[str, Any]) -> dict[str, Any]:
    """
    Mirrors ExecutionTrustRecordCanonicalView.ts's
    canonicalExecutionTrustRecord() field-for-field. A field absent
    from `record` (e.g. no `authorization`) is simply omitted here,
    the same as CanonicalSerializer.ts dropping an undefined key via
    JSON.stringify.
    """

    fields_in_order = (
        "trustRecordId",
        "businessTransactionId",
        "transaction",
        "authorization",
        "overrides",
        "executions",
        "createdAt",
    )

    return {key: record[key] for key in fields_in_order if key in record}


@dataclass(frozen=True)
class OfflineVerificationResult:
    valid: bool
    hash_valid: bool
    legacy_signature_valid: bool
    algorithms_checked: list[str] = field(default_factory=list)
    errors: list[str] = field(default_factory=list)


def verify_execution_trust_record_offline(
    trust_record: dict[str, Any],
    public_keys: dict[str, str],
) -> OfflineVerificationResult:
    """
    `trust_record` is a plain dict, exactly as `json.load()` would
    produce from a Trust Record JSON file or API response body.
    `public_keys` maps keyId -> PEM-encoded public key text.
    """

    errors: list[str] = []
    algorithms_checked: list[str] = []

    canonical = _canonical_execution_trust_record(trust_record)
    digest = Hash(SHA256())
    digest.update(canonical_serialize(canonical))
    expected_hash = digest.finalize().hex()

    hash_valid = expected_hash == trust_record.get("trustRecordHash")

    if not hash_valid:
        errors.append(
            f"trustRecordHash mismatch: expected {expected_hash}, "
            f"got {trust_record.get('trustRecordHash')}."
        )

    signature_field = trust_record.get("signature") or {}
    algorithm = signature_field.get("algorithm")
    key_id = signature_field.get("keyId")
    signature_value = signature_field.get("value")

    legacy_signature_valid = False

    if algorithm not in _SUPPORTED_ALGORITHMS:
        errors.append(f"unsupported algorithm: {algorithm}.")
    elif key_id not in public_keys:
        errors.append(f'no public key supplied for keyId "{key_id}".')
    else:
        algorithms_checked.append(algorithm)

        try:
            public_key = load_pem_public_key(public_keys[key_id].encode("utf-8"))

            if not isinstance(public_key, Ed25519PublicKey):
                raise ValueError("supplied public key is not an Ed25519 key")

            if not isinstance(signature_value, str):
                raise ValueError("signature.value is missing or not a string")

            import base64

            signature_bytes = base64.b64decode(signature_value)
            message = canonical_serialize(canonical)

            _verify_ed25519_with_commitment(public_key, signature_bytes, message)
            legacy_signature_valid = True
        except InvalidSignature:
            errors.append(
                f'signature verification failed for keyId "{key_id}" ({algorithm}).'
            )
        except Exception as error:  # noqa: BLE001 -- reported, not swallowed
            errors.append(f'error verifying keyId "{key_id}" ({algorithm}): {error}')

    if trust_record.get("signatures"):
        errors.append(
            "this record carries a hybrid `signatures` array; ML-DSA-65 "
            "verification is not yet available from Python (see this "
            "module's own docstring) -- verify with the TypeScript "
            "reference implementation (packages/crypto/src/"
            "OfflineVerifier.ts) for a complete hybrid check."
        )

    valid = hash_valid and legacy_signature_valid and not trust_record.get("signatures")

    return OfflineVerificationResult(
        valid=valid,
        hash_valid=hash_valid,
        legacy_signature_valid=legacy_signature_valid,
        algorithms_checked=algorithms_checked,
        errors=errors,
    )


def _canonical_execution_intent(intent: dict[str, Any]) -> dict[str, Any]:
    """
    Mirrors packages/crypto/src/ExecutionIntentCanonicalView.ts's
    canonicalExecutionIntent() field for field (ADR-0012). The signature covers
    exactly these fields, never `intentHash` or `signature`. A field absent
    from `intent` (an optional one that was never set) is omitted, the same as
    CanonicalSerializer.ts dropping an undefined key.
    """

    fields_in_order = (
        "intentId",
        "businessTransactionId",
        "decisionId",
        "authorizationId",
        "policyName",
        "policyVersion",
        "policyContentHash",
        "signalsHash",
        "businessTransactionHash",
        "action",
        "target",
        "submittedBy",
        "grantedCapability",
        "createdAt",
    )

    return {key: intent[key] for key in fields_in_order if key in intent}


def verify_execution_intent_offline(
    intent: dict[str, Any],
    public_keys: dict[str, str],
) -> OfflineVerificationResult:
    """
    Python counterpart to verifyExecutionIntentOffline() in
    packages/crypto/src/OfflineVerifier.ts (ADR-0012). No network call, no
    database, no environment variable: only the intent and the public key(s).

    `intent` is a plain dict, exactly as `json.load()` would produce from the
    `intent` field of GET /execution-intents/:id. `public_keys` maps keyId to
    PEM-encoded public key text.

    A valid result proves the intent was signed by the holder of that key and
    has not been altered. It does NOT prove the action was released, or what its
    result was: an intent is written BEFORE release.
    """

    errors: list[str] = []
    algorithms_checked: list[str] = []

    canonical = _canonical_execution_intent(intent)
    digest = Hash(SHA256())
    digest.update(canonical_serialize(canonical))
    expected_hash = digest.finalize().hex()

    hash_valid = expected_hash == intent.get("intentHash")

    if not hash_valid:
        errors.append(
            f"intentHash mismatch: expected {expected_hash}, "
            f"got {intent.get('intentHash')}."
        )

    signature_field = intent.get("signature") or {}
    algorithm = signature_field.get("algorithm")
    key_id = signature_field.get("keyId")
    signature_value = signature_field.get("value")

    signature_valid = False

    if algorithm not in _SUPPORTED_ALGORITHMS:
        errors.append(f"unsupported algorithm: {algorithm}.")
    elif key_id not in public_keys:
        errors.append(f'no public key supplied for keyId "{key_id}".')
    else:
        algorithms_checked.append(algorithm)

        try:
            public_key = load_pem_public_key(public_keys[key_id].encode("utf-8"))

            if not isinstance(public_key, Ed25519PublicKey):
                raise ValueError("supplied public key is not an Ed25519 key")

            if not isinstance(signature_value, str):
                raise ValueError("signature.value is missing or not a string")

            import base64

            _verify_ed25519_with_commitment(
                public_key,
                base64.b64decode(signature_value),
                canonical_serialize(canonical),
            )
            signature_valid = True
        except InvalidSignature:
            errors.append(
                f'signature verification failed for keyId "{key_id}" ({algorithm}).'
            )
        except Exception as error:  # noqa: BLE001 -- reported, not swallowed
            errors.append(f'error verifying keyId "{key_id}" ({algorithm}): {error}')

    return OfflineVerificationResult(
        valid=hash_valid and signature_valid,
        hash_valid=hash_valid,
        legacy_signature_valid=signature_valid,
        algorithms_checked=algorithms_checked,
        errors=errors,
    )

File: typescript/src/crypto/offline-verifier.ts
/**
 * Offline verification of Execution Trust Records and Execution Intents.
 *
 * TypeScript counterpart of the Python SDK's parmana/crypto/
 * offline_verifier.py, with the same checks and the same result fields in
 * camelCase. No network call, no disk read, no environment variable: only
 * the record and the public keys you pass in. Get a deployment's public key
 * once with `client.publicKey("default")` and keep it.
 *
 * Supports Ed25519, including the large message commitment AWS KMS uses for
 * a canonical form over 4096 bytes (docs/adr/ADR-0010). A record that also
 * carries a hybrid `signatures` array (ML-DSA-65) is reported as not valid
 * here, the same as in the Python SDK, because that algorithm is not
 * verified by the SDKs.
 */

import { createHash, createPublicKey, verify } from "node:crypto";

import { canonicalSerialize } from "./canonical.js";

const SUPPORTED_ALGORITHMS = new Set(["ed25519"]);

const KMS_RAW_MESSAGE_LIMIT_BYTES = 4096;

const COMMITMENT_PREFIX = Buffer.from(
  "PARMANA-ED25519-LARGE-MESSAGE-V1\0",
  "utf8",
);

/**
 * The result of an offline verification.
 */
export interface OfflineVerificationResult {
  /**
   * True only when the hash matches and the signature verifies.
   */
  readonly valid: boolean;

  /**
   * The record's own hash matches a fresh SHA-256 of its canonical form.
   */
  readonly hashValid: boolean;

  /**
   * The Ed25519 signature verifies with the supplied public key.
   */
  readonly legacySignatureValid: boolean;

  readonly algorithmsChecked: readonly string[];

  /**
   * Why verification failed, in plain words. Empty when valid.
   */
  readonly errors: readonly string[];
}

/**
 * Maps a key ID (the record's `signature.keyId`, usually `default`) to that
 * key's PEM public key.
 */
export type PublicKeys = Readonly<Record<string, string>>;

type JsonObject = Readonly<Record<string, unknown>>;

function pick(record: JsonObject, fields: readonly string[]): JsonObject {
  const picked: Record<string, unknown> = {};

  for (const field of fields) {
    if (field in record) {
      picked[field] = record[field];
    }
  }

  return picked;
}

const TRUST_RECORD_FIELDS = [
  "trustRecordId",
  "businessTransactionId",
  "transaction",
  "authorization",
  "overrides",
  "executions",
  "createdAt",
] as const;

const INTENT_FIELDS = [
  "intentId",
  "businessTransactionId",
  "decisionId",
  "authorizationId",
  "policyName",
  "policyVersion",
  "policyContentHash",
  "signalsHash",
  "businessTransactionHash",
  "action",
  "target",
  "submittedBy",
  "grantedCapability",
  "createdAt",
] as const;

function verifyEd25519(
  publicKeyPem: string,
  signature: Buffer,
  message: Uint8Array,
): boolean {
  const publicKey = createPublicKey(publicKeyPem);

  if (publicKey.asymmetricKeyType !== "ed25519") {
    throw new Error("supplied public key is not an Ed25519 key");
  }

  if (verify(null, message, publicKey, signature)) {
    return true;
  }

  // A message over the KMS limit may have been signed as a commitment.
  if (message.length <= KMS_RAW_MESSAGE_LIMIT_BYTES) {
    return false;
  }

  const commitment = Buffer.concat([
    COMMITMENT_PREFIX,
    createHash("sha512").update(message).digest(),
  ]);

  return verify(null, commitment, publicKey, signature);
}

function verifySigned(
  artifact: JsonObject,
  fields: readonly string[],
  hashField: "trustRecordHash" | "intentHash",
  publicKeys: PublicKeys,
): {
  hashValid: boolean;
  signatureValid: boolean;
  algorithmsChecked: string[];
  errors: string[];
} {
  const errors: string[] = [];
  const algorithmsChecked: string[] = [];

  const canonical = canonicalSerialize(pick(artifact, fields));
  const expectedHash = createHash("sha256").update(canonical).digest("hex");
  const hashValid = expectedHash === artifact[hashField];

  if (!hashValid) {
    errors.push(
      `${hashField} mismatch: expected ${expectedHash}, got ${String(artifact[hashField])}.`,
    );
  }

  const signature = (artifact.signature ?? {}) as JsonObject;
  const algorithm = signature.algorithm;
  const keyId = signature.keyId;
  const value = signature.value;

  let signatureValid = false;

  if (typeof algorithm !== "string" || !SUPPORTED_ALGORITHMS.has(algorithm)) {
    errors.push(`unsupported algorithm: ${String(algorithm)}.`);
  } else if (typeof keyId !== "string" || !(keyId in publicKeys)) {
    errors.push(`no public key supplied for keyId "${String(keyId)}".`);
  } else {
    algorithmsChecked.push(algorithm);

    try {
      if (typeof value !== "string") {
        throw new Error("signature.value is missing or not a string");
      }

      signatureValid = verifyEd25519(
        publicKeys[keyId] as string,
        Buffer.from(value, "base64"),
        canonical,
      );

      if (!signatureValid) {
        errors.push(
          `signature verification failed for keyId "${keyId}" (${algorithm}).`,
        );
      }
    } catch (error) {
      errors.push(
        `error verifying keyId "${keyId}" (${algorithm}): ${
          error instanceof Error ? error.message : String(error)
        }`,
      );
    }
  }

  return { hashValid, signatureValid, algorithmsChecked, errors };
}

/**
 * Verifies an Execution Trust Record with only public keys.
 *
 * @param trustRecord The record exactly as `client.trustRecord()` or
 *   `GET /trust-records/{id}` returned it, or as parsed from a JSON file.
 * @param publicKeys Key ID to PEM public key.
 */
export function verifyExecutionTrustRecordOffline(
  trustRecord: unknown,
  publicKeys: PublicKeys,
): OfflineVerificationResult {
  const record = (trustRecord ?? {}) as JsonObject;
  const result = verifySigned(
    record,
    TRUST_RECORD_FIELDS,
    "trustRecordHash",
    publicKeys,
  );

  const hybrid =
    Array.isArray(record.signatures) && record.signatures.length > 0;

  if (hybrid) {
    result.errors.push(
      "this record carries a hybrid `signatures` array; ML-DSA-65 " +
        "verification is not available in the SDKs. Verify it with the " +
        "server's reference implementation (scripts/verify-trust-record.ts).",
    );
  }

  return {
    valid: result.hashValid && result.signatureValid && !hybrid,
    hashValid: result.hashValid,
    legacySignatureValid: result.signatureValid,
    algorithmsChecked: result.algorithmsChecked,
    errors: result.errors,
  };
}

/**
 * Verifies an Execution Intent with only public keys.
 *
 * A valid result proves the intent was signed by the holder of that key and
 * has not been altered. It does not prove the action was released, or what
 * its result was: an intent is written before release.
 *
 * @param intent The `intent` field of `client.executionIntent()` or
 *   `GET /execution-intents/{id}`.
 * @param publicKeys Key ID to PEM public key.
 */
export function verifyExecutionIntentOffline(
  intent: unknown,
  publicKeys: PublicKeys,
): OfflineVerificationResult {
  const result = verifySigned(
    (intent ?? {}) as JsonObject,
    INTENT_FIELDS,
    "intentHash",
    publicKeys,
  );

  return {
    valid: result.hashValid && result.signatureValid,
    hashValid: result.hashValid,
    legacySignatureValid: result.signatureValid,
    algorithmsChecked: result.algorithmsChecked,
    errors: result.errors,
  };
}
